python_modules/dagster/dagster/_core/definitions/asset_out.py

```diff
@@ -117,6 +122,7 @@ def __new__(
                 backfill_policy, "backfill_policy", BackfillPolicy
             ),
             owners=check.opt_sequence_param(owners, "owners", of_type=str),
+            tags=check.opt_mapping_param(tags, "tags", key_type=str, value_type=str),
```

Comment:
Prefer having calls `validate_definition_tags` right next to entry points for cleaner stack traces and the ability to have more specific error messages if appropriate. Extra credit points for capturing that you have validated in the type system.

Reply:
Here's what led me to this:
- People can directly construct `AssetsDefinition`s, so validation at that level seems necessary
- Regexes can be a little bit expensive, so didn't want to cavalierly validate twice (once at the edge and also once inside `AssetsDefinition`)

If we wanted to handle this in the type system, we could introduce a `Tags` class (or `DefinitionTags`?) and accept `Optional[Union[Tags, Mapping[str, str]]]` for all the tags arguments instead of `Optional[Mapping[str, str]]`.  Would that be your preference?

Reply:
Yup def would need to check twice unless you introduce the subtype and do an instanceof check for forgo the regex. But I'm very skeptical that this regex is material to performance, and I think it is a nicer devex to treat as a close to the edge as possible. However take this as a recommendation I don't feel super strongly about it.